## Day 46 Lecture 2 Assignment

In this assignment, we will calculate a similarity matrix using responses to a survey about student life at a university. We will perform clustering on this dataset later on.

In [1]:
%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy.spatial.distance import pdist, squareform

This dataset consists of 35 binary features, each corresponding to a yes/no question that characterizes the student taking the survey.

This dataset contains a large number of features, each corresponding to a survey question. The feature name summarizes the survey question, so we will not list them all out here.

Load the dataset.

In [2]:
# answer goes here
url = 'https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/Data%20Sets%20Clustering/student_life_survey.csv'

survey_df = pd.read_csv(url)

survey_df.info()
survey_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2958 entries, 0 to 2957
Data columns (total 36 columns):
 #   Column                                                                                                                      Non-Null Count  Dtype
---  ------                                                                                                                      --------------  -----
 0   Q2-Participated in Societies and Interest Groups                                                                            2958 non-null   int64
 1   Q2-Participated in Clubs                                                                                                    2958 non-null   int64
 2   Q2-Participated in Halls, JCRCs and/or Residential College CSCs                                                             2958 non-null   int64
 3   Q2-Participated in University organised events                                                                              2958 non-

Q2-Participated in Societies and Interest Groups  ...  response_id
0                                                 0  ...            1
1                                                 0  ...            2
2                                                 0  ...            3
3                                                 1  ...            4
4                                                 1  ...            5

[5 rows x 36 columns]

For our analysis, we will focus on a specific subset of the survey that is focused on stress. These questions all begin with the string 'Q5'. Filter the columns that meet this criteria (should be 10 in total).

In [3]:
# answer goes here
stress_df = survey_df.filter(like='Q5', axis=1)

stress_df

Q5-Stressed about Adjustment issues  ...  Q5-Stressed about Others
0                                       0  ...                         0
1                                       0  ...                         0
2                                       0  ...                         0
3                                       1  ...                         0
4                                       0  ...                         0
...                                   ...  ...                       ...
2953                                    0  ...                         0
2954                                    0  ...                         0
2955                                    0  ...                         0
2956                                    0  ...                         0
2957                                    0  ...                         0

[2958 rows x 10 columns]

We'll assume for the next step that a pair of negative values (i.e. both responses are "no") is as informative as a pair of positive values. Compute the full distance/dissimilarity matrix for the survey data using matching/hamming distance.

The pdist function is a simple way to accomplish this goal, and the squareform function will transform the output of the pdist function into a tabular format that can be convert to a dataframe more easily.

In [4]:
# answer goes here
ham_matrix = pd.DataFrame(squareform(pdist(stress_df, metric ='hamming')), 
                          index=stress_df.index, columns=stress_df.index)
np.fill_diagonal(ham_matrix.values, np.nan)
ham_matrix

0     1     2     3     4     5     ...  2952  2953  2954  2955  2956  2957
0      NaN   0.0   0.1   0.6   0.3   0.1  ...   0.3   0.2   0.0   0.0   0.2   0.2
1      0.0   NaN   0.1   0.6   0.3   0.1  ...   0.3   0.2   0.0   0.0   0.2   0.2
2      0.1   0.1   NaN   0.7   0.4   0.2  ...   0.4   0.3   0.1   0.1   0.1   0.1
3      0.6   0.6   0.7   NaN   0.5   0.5  ...   0.3   0.6   0.6   0.6   0.6   0.8
4      0.3   0.3   0.4   0.5   NaN   0.2  ...   0.2   0.3   0.3   0.3   0.5   0.3
...    ...   ...   ...   ...   ...   ...  ...   ...   ...   ...   ...   ...   ...
2953   0.2   0.2   0.3   0.6   0.3   0.3  ...   0.5   NaN   0.2   0.2   0.4   0.2
2954   0.0   0.0   0.1   0.6   0.3   0.1  ...   0.3   0.2   NaN   0.0   0.2   0.2
2955   0.0   0.0   0.1   0.6   0.3   0.1  ...   0.3   0.2   0.0   NaN   0.2   0.2
2956   0.2   0.2   0.1   0.6   0.5   0.3  ...   0.5   0.4   0.2   0.2   NaN   0.2
2957   0.2   0.2   0.1   0.8   0.3   0.3  ...   0.5   0.2   0.2   0.2   0.2   NaN

[2958 rows x 2958 columns]

Print the answers to the following three questions:

1. How many pairs of students are there in our dataset? (Remember to ignore a student pairing with themselves, and avoid double-counting - '3-5' is the same as '5-3'.) 
2. Of these pairs, how many gave identical answers? 
3. How many pairs had no answers in common?

Tip: it may help to first convert the NxN similarity matrix into a table with three columns: response_id_1, response_id_2, and the similarity between them. The new table should have NxN rows and 3 columns.

In [5]:
# n*(n-1)/2
2958*(2958-1)/2

4373403.0

In [6]:
ham_count = ham_matrix.apply(pd.value_counts).T

ham_count

0.0    0.1    0.2     0.3    0.4    0.5    0.6    0.7    0.8   0.9  1.0
0     341.0  717.0  804.0   525.0  294.0  155.0   71.0   29.0   21.0   NaN  NaN
1     341.0  717.0  804.0   525.0  294.0  155.0   71.0   29.0   21.0   NaN  NaN
2     260.0  696.0  763.0   652.0  329.0  154.0   72.0   30.0    1.0   NaN  NaN
3       1.0   16.0   68.0   164.0  348.0  625.0  912.0  610.0  202.0  11.0  NaN
4      12.0  105.0  491.0  1036.0  841.0  381.0   81.0   10.0    NaN   NaN  NaN
...     ...    ...    ...     ...    ...    ...    ...    ...    ...   ...  ...
2953   21.0  206.0  757.0   910.0  640.0  286.0  107.0   23.0    6.0   1.0  NaN
2954  341.0  717.0  804.0   525.0  294.0  155.0   71.0   29.0   21.0   NaN  NaN
2955  341.0  717.0  804.0   525.0  294.0  155.0   71.0   29.0   21.0   NaN  NaN
2956   18.0  334.0  775.0   790.0  632.0  279.0  104.0   23.0    2.0   NaN  NaN
2957  124.0  575.0  891.0   647.0  433.0  186.0   81.0   18.0    2.0   NaN  NaN

[2958 rows x 11 columns]

In [7]:
ham_count[0.0].sum()/2

136732.0

In [8]:
ham_count[1.0].sum()/2

291.0

1. How many pairs of students are there in our dataset? 
> 4373403
2. Of these pairs, how many gave identical answers? 
> 136732
3. How many pairs had no answers in common?
> 291

For each student, calculate their average similarity with all other students. Then print the following:

1. On average, which student (i.e. which response_id) had the most similar responses to the survey population as a whole? What was the average similarity for this student, and what were their answers?
2. Repeat part 1, but for the student with the least similar responses to the survey population as a whole.
3. What do we notice about these students? What broader inferences can we make about the student population as a whole?

In [9]:
# answer goes here
avg_sim = pd.DataFrame({'response_id':survey_df['response_id'], 'avg_similarity_ham': ham_matrix.mean()})

avg_sim = pd.concat([avg_sim, stress_df],1)

In [10]:
avg_sim.sort_values(by='avg_similarity_ham', ascending=True).head()

response_id  ...  Q5-Stressed about Others
0               1  ...                         0
440           441  ...                         0
1865         1866  ...                         0
1869         1870  ...                         0
433           434  ...                         0

[5 rows x 12 columns]

In [11]:
avg_sim.loc[avg_sim['avg_similarity_ham']==avg_sim['avg_similarity_ham'].min()]

response_id  ...  Q5-Stressed about Others
0               1  ...                         0
1               2  ...                         0
23             24  ...                         0
39             40  ...                         0
41             42  ...                         0
...           ...  ...                       ...
2936         2937  ...                         0
2940         2941  ...                         0
2942         2943  ...                         0
2954         2955  ...                         0
2955         2956  ...                         0

[342 rows x 12 columns]

In [12]:
avg_sim.sort_values(by='avg_similarity_ham', ascending=False).head(10)

response_id  ...  Q5-Stressed about Others
2899         2900  ...                         1
2416         2417  ...                         1
923           924  ...                         0
1338         1339  ...                         0
1605         1606  ...                         0
245           246  ...                         0
2041         2042  ...                         0
1578         1579  ...                         0
1796         1797  ...                         0
263           264  ...                         0

[10 rows x 12 columns]

In [13]:
avg_sim.loc[avg_sim['avg_similarity_ham']>=0.593777]

response_id  ...  Q5-Stressed about Others
245           246  ...                         0
263           264  ...                         0
496           497  ...                         0
752           753  ...                         0
812           813  ...                         0
866           867  ...                         0
878           879  ...                         0
923           924  ...                         0
941           942  ...                         0
1125         1126  ...                         0
1282         1283  ...                         0
1291         1292  ...                         0
1338         1339  ...                         0
1472         1473  ...                         0
1578         1579  ...                         0
1605         1606  ...                         0
1796         1797  ...                         0
2041         2042  ...                         0
2416         2417  ...                         1
2706         2707  ...                         0
2724         2725  ...                         0
2899         2900  ...                         1

[22 rows x 12 columns]

1. On average, which student (i.e. which response_id) had the most similar responses to the survey population as a whole? What was the average similarity for this student, and what were their answers?
> 342 students had the most similar average response of 0.224822. These students were only stressed about 'Academic Issues'.
2. Repeat part 1, but for the student with the least similar responses to the survey population as a whole.
> Generally, the least similar average response were by students who were stressed by almost all factors besides 'Others'.
3. What do we notice about these students? What broader inferences can we make about the student population as a whole?
> Most students aren't extremely stressed. Most students only have a few issues of stress, Academic being the most common. 